modnet/fit_genetic.py

```diff
@@ -0,0 +1,246 @@
+            monitor="loss",
+            factor=0.5,
+            patience=20,
+            verbose=0,
+            mode="auto",
+            min_delta=0,
+        )
+        es = keras.callbacks.EarlyStopping(
+            monitor="loss",
+            min_delta=0.001,
+            patience=300,
+            verbose=0,
+            mode="auto",
+            baseline=None,
+            restore_best_weights=True,
+        )
+        callbacks = [rlr, es]
+        for w in self.pop:
+            modnet_model = MODNetModel([[['BV_Ea']]], {'BV_Ea':1}, n_feat=w[0], num_neurons=[[int(w[1])],[int(w[1]*w[2])],[int(w[1]*w[2]*w[3])],[int(w[1]*w[2]*w[3]*w[4])]], act=w[5])
+            try:
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 2*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 4*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                modnet_model.fit(X_train,val_fraction=0, val_key='BV_Ea',loss=w[6], lr=w[8], epochs = 250, batch_size = 8*w[9], xscale=w[7], callbacks=callbacks, verbose=0)
+                f = mse(modnet_model.predict(X_val),y_val)
+                print('MSE = ', f)
+                self.fitness.append([f, modnet_model, w])
+            except:
+                 pass
+        return self.fitness
+
+
+    def gen_alg(
+        self,
+        X_train: MODData,
+        y_train: pd.DataFrame,
+        X_val: MODData,
+        y_val: pd.DataFrame,
+        size_pop: int,
+        num_epochs: int,
+        prob_mut: float = 0.8
+        )->None:
+
+        """Selects the best individual (the model with the best parameters) for the next generation. The selection is based on a minimisation of the MSE on the validation set.
+
+        Parameters:
+            X_train: Input data of the training set.
+            y_train: Target values of the training set.
+            X_val: Input data of the validation set.
+            y_val: Target values of the validation set.
+            size_pop: Size of the population per generation.
+            num_epochs: Number of generations.
+            prob_mut: Probability the mutation occurs.
+        """
+
+        print('Generation number 0')
+        pop = self.initialization_population(size_pop)
+        fitness = self.function_fitness(pop,  X_train, y_train, X_val, y_val)
+        pop_fitness_sort = np.array(list(sorted(fitness,key=lambda x: x[0])))
+        for j in range(0, num_epochs):
+            print('Generation number ', j+1)
+            length = len(pop_fitness_sort)
+            #select parents
+            parent_1 = pop_fitness_sort[:,2][:length//2]
+            parent_2 = pop_fitness_sort[:,2][length//2:]
```

Comment:
You should probably select parents randomly weighted by their fitness, rather than by a rank order.

Reply:
Just to expand on what I said in the meeting:

- Fitness should be rescaled to [0, 1] and used to weight the choice of parents. You can do this with e.g. `mother, father = random.choices(population, weights=scaled_pop_fitness, k=2)` or similar.
- It will be beneficial to apply some variant of a logistic function to the fitness (e.g. `(tanh(scaled_pop_fitness) + 1)/2` to keep it between [0, 1]) such that low fitness models are more heavily penalised, and those close to the best model are not so heavily penalised as they might still provide fruitful search directions.